src/cargo/ops/cargo_update.rs

```diff
@@ -165,7 +182,15 @@ pub fn update_lockfile(ws: &Workspace<'_>, opts: &UpdateOptions<'_>) -> CargoRes
         .filter(|s| !s.is_registry())
         .collect();
 
-    let keep = |p: &PackageId| !to_avoid_sources.contains(&p.source_id()) && !to_avoid.contains(p);
+    let breaking_precise_upgrades = opts.precise.is_some() && !upgrades.is_empty();
+    let breaking_mode = opts.breaking || breaking_precise_upgrades;
+
+    let keep = |p: &PackageId| {
+        (!to_avoid_sources.contains(&p.source_id()) && !to_avoid.contains(p))
+            // In case of `--breaking` or precise upgrades, we want to keep all
+            // packages unchanged that didn't get upgraded.
+                || (breaking_mode && !upgrades.contains_key(&(p.name().to_string(), p.source_id())))
+    };
```

Comment:
@torhovland I was very intentionally trying to avoid changing this in past PRs.  To go back and change it requires a good explanation.  Even better if you can also split this into its own commit or PR

@Eh2406 I'm very cautious of messing with `keep` and would appreciate your input on tweaking it for this case

Reply:
`keep` also makes my head spin. I feel like it's something about the triple negative structure, we update the ones that are not `keep`, we `keep` the ones that are not `avoid`, we `avoid` the things that asked to be updated. But I've never found it clearer naming structure, so maybe that's not it. It's also trying to do a linear prediction of resolution which feels like an approximation.

In specific, why does this need to be a separate lookup as opposed to adding the matching packages to `to_avoid`?

Also, this callback is run a lot on no-op builds, so very small performance problems end up being a high percentage of runtime. I suspect the `to_string` is going to end up being expensive. Can the `upgrades` map be `InternedStrings` or `&str`?

Reply:
I'm pulling this out into a separate PR: https://github.com/rust-lang/cargo/pull/14259

Reply:
I assume with that pulled out this PR will be updated so we can get it merged without #14259 so this doesn't get blocked on that decision?

Reply:
I don't like the idea of not building on #14259. I suggest we rather let this PR depend on #14259 and leave it as draft for now. Here's why:

- Some of the duplicated existing precise tests with the unstable feature enabled are no longer passing. In other words, without #14259 we will be changing the behaviour of the non-breaking precise update.

- Some invocations are now silently finishing rather than reporting an error. For example:

```
[ERROR] package ID specification `incompatible@2.0.0` did not match any packages
Did you mean one of these?

  incompatible@0.3.1
```

- The output is generally no longer consistent with the non-breaking update. For example, without #14259 the output changes like this:

```
-[UPDATING] incompatible v0.1.0 -> v0.2.0
-[NOTE] pass `--verbose` to see 1 unchanged dependencies behind latest
+[LOCKING] 1 package to latest compatible version
+[UPDATING] incompatible v0.1.0 -> v0.2.0 (latest: v0.2.1)
```

Reply:
> I don't like the idea of not building on https://github.com/rust-lang/cargo/pull/14259. I suggest we rather let this PR depend on https://github.com/rust-lang/cargo/pull/14259 and leave it as draft for now.

That PR has behavior changes that have not been agreed to, internal and user-facing.  We should not block progress forward on something on changes like that.  If there is some intermediate change that side steps the controversial parts that unblocks `--precise`, then let's go for that.

Reply:
Is #14259 still controversial? It no longer changes `keep`. I hesitate moving forward with this PR without merging #14259 first, particularly because of the first bullet point above (breaking compatibility with the existing precise update).